### PR TITLE
[FIX][I] #251 Fix EditorManager.saveEditors(...)

### PR DESCRIPTION
**Pull request by tobous**
EditorManager.saveEditors(IProject) now uses the EditorPool to determine
which editors are open instead of holding a separate list. This separate
list was superfluous and was leading to issues when trying to save the
open Editors, as the list was not updated correctly, leading to
inconsistencies when starting a session while having an unsaved file
open.

Subsequently removes the hashmap used to store the open editors and
adjusts other methods using this second editor store.

### CODE REVIEW

**Comment by tobous** [comment]
This class was initially copied from Saros/E. This logic is a remnant of this as it also exists in Saros/E. I did some digging and it was introduced that way in 2006 with https://github.com/saros-project/saros/commit/f0e78dd2fb6384aa71c0afdd1f9794bc369c9f1d. (The initial commit starting this repos https://github.com/saros-project/saros/commit/155cdba93ae04cf2b3ffaee86b5196bcad072631 did not have this code, but all files of this commit were removed with the next commit and a new version of Saros was introduced with the commit mentioned above.) The EditorPool was already present at that point, but maybe it did different things? I don't know.

I don't know why this divide exists, either. I just though it was an oversight while implementing Saros/I (or rather while adjusting the class for IntelliJ).

From my point of view, the initial implementation of `saveEditors(IProject)` does not make much sense either, as it does not match the [`IEditorStateManager` javadoc](https://github.com/saros-project/saros/blob/9fc29671a436b5cbd9d72ddc657abc425e3900cc/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/editor/IEditorManager.java#L65):
It tries to also save all editors contained in the `UserEditorStateManager`. This is somewhat pointless as the `UserEditorStateManager` only stores remote editors, meaning there does not have to be a local editor for the file. Furthermore, the javadoc of `IEditorStateManager` explicitly states that the method is supposed to save *local* editors.

@srossbach @fzieris Do you know why the EditorManager is holding a separate list of all open editors, even though it also holds the EditorPool? Do you know what exactly `IEditorManager.saveEditors(IProject)` is supposed to do?

**Comment by srossbach** [comment]
The EditorPool is the collection of all editors open by the local. It is used to keep track and install various listeners. On the other hand UserEditorStateManager keeps track on the "remote editors". If you do not know the internal you might think, why I need the UserEditorStateManager ? It is needed because text edits are applied through the editor, even it is not opened on the local side. So it does INDEED only save / flush the local editors (I did not check if the save is broadcasted but I guess it is not)

As for your other question, I really cannot tell you why there is a EditorPool and also a separate list without digging through the EditorManager code.

As for IEditorManager.saveEditors(IProject) after I call this method for a given / or all projects -> all dirty / cached editor contents for files that are part of the sharing have to successfully written to the underlying storage / filesystem -> If I read the file through the filesystem it will have the same content as if I would access the editor content through the given editor provider.

**Comment by tobous** [comment]
Ok, then I interpreted the javadoc to pedantically. The used wording does not really apply to Intellij as the Editors don't hold the content of the file, they are only the interface the user interacts with. The actual content is stored in [Documents](http://www.jetbrains.org/intellij/sdk/docs/basics/architectural_overview/documents.html). There can't be any unsaved "editors" that don't exist locally.

But that is not the point. As already mentioned in #255, I don't think flushing to disk instead of reading from the cache directly is not the right way to go. But I also think that saving now and again is a good idea, just in case the application crashed or there is a problem with the caching logic. I will adjust this logic to save all "open" editors.

**Comment by srossbach** [comment]
The Eclipse stuff does not differ that much from IntelliJ

https://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Fguide%2Feditors_documents.htm

Basically you have an IEditorpart (the visible stuff) and an IEditorInput where the content is fetched and stored. Between those two interfaces you have a document provider that gives you access to a document instance that is hold in memory containing the editor contents.

I really do not know why everything is prefixed with Editor in our logic as some logic operates on documents but the logic was written long ago.

I do not know if it makes sense to split things, so having an EditorManager and a DocumentManager, I guess this will be confusing as well.

**Comment by tobous** [comment]
@srossbach Do you still have any issues with the PR?

I think splitting the EditorManager in general is a very good idea. It is way to big and does way to much. (For IntelliJ) It executes TextEdit-, Editor, TextSelection-, and Viewport-Activities. Furthermore, it dispatches some EditorActivities, provides some editor utility functions, and it provides functions to generate and dispatch the above mentioned activities (even though these functions are never used inside the class itself).

Splitting of the document stuff would be easy-ish as it (almost) exclusively concerns the part dealing with TextEditActivities. The only problem is that saving also concerns the documents, as this can't be done with an editor. You can easily get the document for an editor, but this would mix concerns again. A solution would be to also listen for SAVED EditorActivities in the document part.

But, as I have already said (too) many times before: I don't think this hurts enough (at least for IntelliJ) for me to deal with it right now. There are still many parts of the basic Saros functionality that are broken or missing in Saros/I.

**Comment by tobous** [comment]
@srossbach Do you still have any issues with this PR? Otherwise, I would appreciate you to approving it?

File: de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/LocalEditorHandler.java [hunk 1]
```diff
@@ -196,17 +197,41 @@ public void removeEditor(@NotNull SPath path){
...
-     * @param path
+     * @param path the path for the document to save
      */
-    public void saveFile(SPath path) {
-        Document doc = editorPool.getDocument(path);
-        if (doc != null) {
-            projectAPI.saveDocument(doc);
-        } else {
-            LOG.warn("Document does not exist: " + path);
+    public void saveFile(
```
Comment (srossbach) [inline de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/LocalEditorHandler.java]: Why not saveDocument ?

Comment (tobous) [reply de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/LocalEditorHandler.java]: Done.

**Pull request merged**